packages/lit-html/src/directives/guard.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  directive,
+  Directive,
+  noChange,
+  nothing,
+  Part
+} from '../lit-html.js';
+
+class Guard extends Directive {
+  previousValue: unknown = nothing;
+
+  render(_value: unknown, f: () => unknown) {
+    return f();
+  }
+
+  update(_part: Part, [value, f]: Parameters<this['render']>) {
+    console.log('update', value);
+    if (Array.isArray(value)) {
+      // Dirty-check arrays by item
+      if (Array.isArray(this.previousValue) &&
+          this.previousValue.length === value.length &&
+          value.every((v, i) => v === (this.previousValue as Array<unknown>)[i])) {
+        console.log('A');
+        return noChange;
+      }
+    } else if (
+        this.previousValue === value &&
+        (value !== undefined || this.previousValue !== nothing)) {
```

Comment:
Can you comment why this is necessary?

Reply:
Yeah, the checks other than the equality check seem unnecessary?

Reply:
What should happen if you explicitly give a value of `noChange`?

Reply:
This was a vestige from the old Directive API where we didn't have a constructor. I removed it.